### ISSUE TITLE
Custom expectation stops parser

### ISSUE THREAD

**Issue by glimow** [pegjs/pegjs#575]
### Issue type

**Question:** 

### Description
Hello guys,
I am trying to do dynamic expectation based on parser's given options.
`dims` and `dateDims` are defined in the initializer

```
Dimension
	= _ field:Field {
		if(!dims.includes(field)){
			expected(dims.join('|'))
		};
    	return {
           // some value
        };
    }

DateDimension
	= _ field:Field _ scale: Scale? {
		if(!dateDims.includes(field)){
			expected(dateDims.join("|"))
		};
    	return {
	// some value
        };
    }
```
But when I am trying to parse this expression:

```(Dimension/DateDimension)```

With a DateDimension, it fails and gives only `Dimensions`'s expectations. Am I doing something wrong ? How to make the parser to try each rules before throwing an error ?

**Comment by pjmolina** [comment]
Hi.
PEG algorithm chooses the first production if match and will not check for the second one.
Just rewrite your rule as: `DateDimension / Dimension`   (more specific first) due to [#573](https://github.com/pegjs/pegjs/issues/573)

**Comment by glimow** [comment]
Thanks for your answer
Unfortunately, putting the more specific at first does not work either because the scale being optional, it matches as well.
Is it possible to dynamically consider the match failed ? Meaning the same logic as `field:Field &{ !!datefield[field] }` but with custom expected values *if and only if no other rules are parseable* ?

**Comment by pjmolina** [comment]
I see. Didn't notice the second rule is identical to the same one, but with an optional part.
Then, you can parse it, for example, in the following way:

```js
GenericDimension
	= _ field:Field _ scale:Scale? 
        {
           if (scale) {
               return DateDimensionAST....
           } else {
               return DimensionAST....
           }
        }
```

Note that scale will be null if the Scale is not present, and then, build different trees based on that.

**Comment by glimow** [comment]
That's what I have done at first, but my main goal here is to make an efficient autocompletion feature. Meaning I want that
- if the dimension's field is a date one, the parser fails asking for a `scale` 
- if not, the parser does not expect a date 
In a more generic way, I wonder if this is possible to use `! { predicate }` while giving the parser correct expectations

**Comment by pjmolina** [comment]
Yes, from [docs](https://pegjs.org/documentation): you can use `& { predicate }` and  `! { predicate }` forms to make rules fail depending on predicates. 
Just remember how PEG works: first rule matching, will ignore next ones when evaluating a _'prioritized choice with backtracking'_: `opt1 / opt2`

**Comment by glimow** [comment]
Well, I can use &{ predicate } to make the rule fail, but this way I would have an `expected: "rulename"` in expectations. What I want is to dynamically compute expectations to send user a feedback about what the predicate expects
